### ISSUE TITLE
Make Index.add_documents compatible with other list-like types

### ISSUE THREAD

**Issue by AlexAltea** [AlexAltea/milli-py#1]
E.g. tuples, dict_values.

Otherwise there might be performance loss during to needless conversions.

**Comment by mardix** [comment]
Couldn't the tuple or the other list types be converted to list ([{...}, ...]) before adding the documents Index.add_documents?

**Comment by AlexAltea** [comment]
Yes, you can always do that, but it might have some performance impact. Imagine you want to index a million documents which happen to be a `tuple`, or `dict_values`.

Doing `list(...)` causes everything to be copied from the original tuple into a *new* list, and *then* processed by Milli.

What I'm trying to do is have Milli's `DocumentBatchBuilder` receive the documents straight from the tuple/dict_values.

**Comment by mardix** [comment]
ok. make sense.

**Comment by AlexAltea** [comment]
Done on 4337774448a6ba04ba934f8b505e1e8f0c21c0cb. Minor performance improvement for `list` and added support for tuples, dict_values and generators with comparable performance according to [`bench.py`](https://github.com/AlexAltea/milli-py/blob/master/benchmarks/bench.py).

| variant | impl | time | peak RSS
| -- | -- | -- | --
| list | old | 9.084 s | 686 MB
| list | new | 8.993 s | 620 MB
| tuple | new | 9.019 s | 619 MB
| dict_values | new | 9.078 s | 632 MB
| generator | new | 9.608 s | 624 MB

This will be available in the coming milli-py version v1.11.1.
